src/lib/ares_process.c

```diff
@@ -816,6 +854,83 @@ static struct server_state *ares__random_server(ares_channel_t *channel)
   return NULL;
 }
 
+/* Pick a server from the list with failover behavior.
+ *
+ * We default to using the first server in the sorted list of servers. That is
+ * the server with the lowest number of consecutive failures and then the
+ * highest priority server (by idx) if there is a draw.
+ *
+ * However, this by itself has two issues:
+ * (1) If a server temporarily goes down and hits some failures, then that
+ *     server will never be retried until all other servers hit the same number
+ *     of failures. This may prevent the server from being retried for a long
+ *     time.
+ * (2) If all servers are failed, then we will continue to select the
+ *     server with the lowest number of consecutive failures even though it may
+ *     be more broken than other servers later in the list.
+ *
+ * For issue (1), with some probability we select a failed server to retry
+ * instead.
+ *
+ * For issue (2), if all servers in the list have a failure, then we fall back
+ * to selecting a random server.
+ */
+static struct server_state *ares__failover_server(ares_channel_t *channel)
+{
+  struct server_state *first_server;
+  struct server_state *last_server;
+  struct timeval       now;
+  unsigned char        c;
+
+  first_server = ares__slist_first_val(channel->servers);
+  last_server  = ares__slist_last_val(channel->servers);
+
+  /* Defensive code against no servers being available on the channel. */
+  if (first_server == NULL) {
+    return NULL;
+  }
+
+  /* If no servers have failures, then prefer the first server in the list. */
+  if (last_server != NULL && last_server->consec_failures == 0) {
+    return first_server;
+  }
+
+  /* If all servers have failures then there is not much reason to prefer one
+   * over another, so select randomly.
+   */
+  if (first_server->consec_failures >= SERVER_ALL_FAILED_THRESHOLD) {
```

Comment:
Think this `if` statement doesn't do anything useful.  I'd remove this code block and let the rest of the code do its thing.  It will auto re-sort based on number of consecutive failures, makes the sequence more predictable as well which makes test cases easier to write and validate.

Reply:
Removed, thanks.